cli/src/cmd/cast/create2.rs

```diff
@@ -0,0 +1,150 @@
+    #[clap(
+        long,
+        short,
+        help = "Prefix for the contract address.",
+        required_unless_present_any = &["ends-with", "matching"],
+        value_name = "HEX"
+    )]
+    starts_with: Option<String>,
+    #[clap(long, short, help = "Suffix for the contract address.", value_name = "HEX")]
+    ends_with: Option<String>,
+    #[clap(long, short, help = "Sequence that the address has to match", value_name = "HEX")]
+    matching: Option<String>,
+    #[clap(short, long)]
+    case_sensitive: bool,
+    #[clap(
+        short,
+        long,
+        help = "Address of the contract deployer.",
+        default_value = "0x4e59b44847b379578588920ca78fbf26c0b4956c",
+        value_name = "ADDRESS"
+    )]
+    deployer: Address,
+    #[clap(
+        short,
+        long,
+        help = "Init code of the contract to be deployed.",
+        value_name = "HEX",
+        default_value = ""
+    )]
+    init_code: String,
+}
+
+impl Cmd for Create2Args {
+    type Output = ();
+
+    fn run(self) -> eyre::Result<Self::Output> {
+        let Create2Args { starts_with, ends_with, matching, case_sensitive, deployer, init_code } =
+            self;
+        Create2Args::generate_address(
+            starts_with,
+            ends_with,
+            matching,
+            case_sensitive,
+            deployer,
+            init_code,
+        )
+    }
+}
+
+impl Create2Args {
+    fn generate_address(
+        starts_with: Option<String>,
+        ends_with: Option<String>,
+        matching: Option<String>,
+        case_sensitive: bool,
+        deployer: Address,
+        init_code: String,
+    ) -> Result<()> {
+        let mut regexs = vec![];
+
+        if let Some(mut matches) = matching {
+            if starts_with.is_some() || ends_with.is_some() {
+                eyre::bail!("Either use --matching or --starts/ends-with");
+            }
+
+            if matches.len() != 40 {
+                if !(matches.len() == 42 && matches.starts_with("0x")) {
+                    eyre::bail!("Please provide a 40 characters long sequence for matching");
+                }
+                matches = matches.strip_prefix("0x").unwrap().to_string();
+            }
+
+            hex::decode(&matches.replace('X', "0")).wrap_err("invalid matching hex provided")?;
+            regexs.push(matches.replace('X', "."));
+        }
+
+        if let Some(prefix) = starts_with {
+            let pad_width = prefix.len() + prefix.len() % 2;
+            hex::decode(format!("{:0>width$}", prefix, width = pad_width))
+                .wrap_err("invalid prefix hex provided")?;
+            regexs.push(format!(r"^{}", prefix));
+        }
+        if let Some(suffix) = ends_with {
+            let pad_width = suffix.len() + suffix.len() % 2;
+            hex::decode(format!("{:0>width$}", suffix, width = pad_width))
+                .wrap_err("invalid suffix hex provided")?;
+            regexs.push(format!(r"{}$", suffix));
+        }
+
```

Comment:
we just made changes to the `cast vanity` command, which should affect this as well

See https://github.com/foundry-rs/foundry/pull/3285

and https://github.com/foundry-rs/foundry/blob/master/cli/src/cmd/cast/wallet/vanity.rs

we should be able to merge these